src/theory/theory_model_builder.cpp

```diff
@@ -26,10 +26,99 @@ using namespace CVC4::context;
 namespace CVC4 {
 namespace theory {
 
+void TheoryEngineModelBuilder::Assigner::initialize(
+    TypeNode tn, TypeEnumeratorProperties* tep, const std::vector<Node>& aes)
+{
+  d_te.reset(new TypeEnumerator(tn, tep));
+  d_assignExcSet.insert(d_assignExcSet.end(), aes.begin(), aes.end());
+}
+
+Node TheoryEngineModelBuilder::Assigner::getNextAssignment()
+{
+  Assert(d_te != nullptr);
+  Node n;
+  bool success = false;
+  TypeEnumerator& te = *d_te;
+  // check if we have run out of elements
+  if (te.isFinished())
+  {
+    Assert(false);
+    return Node::null();
+  }
+  // must increment until we find one that is not in the assignment
+  // exclusion set
+  do
+  {
+    n = *te;
+    success = std::find(d_assignExcSet.begin(), d_assignExcSet.end(), n)
+              == d_assignExcSet.end();
+    // increment regardless of fail or succeed, to set up the next value
+    ++te;
+  } while (!success);
+  return n;
+}
+
 TheoryEngineModelBuilder::TheoryEngineModelBuilder(TheoryEngine* te) : d_te(te)
 {
 }
 
+Node TheoryEngineModelBuilder::evaluateEqc(TheoryModel* m, TNode r)
+{
+  eq::EqClassIterator eqc_i = eq::EqClassIterator(r, m->d_equalityEngine);
+  for (; !eqc_i.isFinished(); ++eqc_i)
+  {
+    Node n = *eqc_i;
+    Trace("model-builder-debug") << "Look at term : " << n << std::endl;
+    if (!isAssignable(n))
+    {
+      Trace("model-builder-debug") << "...try to normalize" << std::endl;
+      Node normalized = normalize(m, n, true);
+      if (normalized.isConst())
+      {
+        return normalized;
+      }
+    }
+  }
+  return Node::null();
+}
+
+bool TheoryEngineModelBuilder::isAssignerActive(TheoryModel* tm, Assigner& a)
+{
+  if (a.d_isActive)
+  {
+    return true;
+  }
+  std::vector<Node>& eset = a.d_assignExcSet;
+  std::map<Node, Node>::iterator it;
+  for (unsigned i = 0, size = eset.size(); i < size; i++)
```

Comment:
We can't since we update at an index on line 116.

Reply:
It should be possible to just do it with `for (Node& er : eset)`.

Reply:
This is pretty unintuitive to me. I'm actually surprised this works.  Since Node is immutable, I'm not sure why modifying the `er` reference in this case would modify `eset`?

Reply:
It works because it is a reference to the element in the `eset` vector (and not a copy of it). So assigning something to it is equivalent to `eset[i] = ...`. Since we are not really using references to `Node`s, maybe we should just leave it as it is though.